flink-ml-core/src/main/java/org/apache/flink/ml/linalg/typeinfo/DenseVectorSerializer.java

```diff
@@ -20,29 +20,36 @@
 package org.apache.flink.ml.linalg.typeinfo;
 
 import org.apache.flink.api.common.typeutils.SimpleTypeSerializerSnapshot;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
 import org.apache.flink.api.common.typeutils.TypeSerializerSnapshot;
-import org.apache.flink.api.common.typeutils.base.TypeSerializerSingleton;
 import org.apache.flink.core.memory.DataInputView;
 import org.apache.flink.core.memory.DataOutputView;
 import org.apache.flink.ml.linalg.DenseVector;
+import org.apache.flink.ml.util.Bits;
 
 import java.io.IOException;
 import java.util.Arrays;
+import java.util.Objects;
 
 /** Specialized serializer for {@link DenseVector}. */
-public final class DenseVectorSerializer extends TypeSerializerSingleton<DenseVector> {
+public final class DenseVectorSerializer extends TypeSerializer<DenseVector> {
 
     private static final long serialVersionUID = 1L;
 
     private static final double[] EMPTY = new double[0];
 
-    public static final DenseVectorSerializer INSTANCE = new DenseVectorSerializer();
+    private final byte[] buf = new byte[1024];
```

Comment:
Instead of add buffer inside serializers and making those serializers non-singleton, would it be simpler to just let caller use `BufferedOutputStream`?

For example, `NaiveBayesModelData::ModelDataEncoder::encode()` can do this:

```
DataOutputViewStreamWrapper outputViewStreamWrapper =
        new DataOutputViewStreamWrapper(new BufferedOutputStream(outputStream));
```

Reply:
According to offline discussion, BufferedOutputStream cannot avoid it that there will be a lot of small byte array read/write, and still have performance issues. We'll still use the internal-byte-buffer practice.